cmd/server/assets/codes/bulkissue.html

```diff
@@ -0,0 +1,289 @@
+      let $progress = $('#progress');
+      let $retryCode = $('#retry-code');
+      let $rememberCode = $('#remember-code');
+      let $newCode = $('#new-code');
+      let $startAt = $('#new-code');
+
+      let randomString = getCookie("retryCode");
+      if (randomString == "") {
+        randomString = genCode();
+      } else {
+        $rememberCode.prop("checked", true);
+      }
+      $retryCode.val(randomString);
+
+      let totalCreated = 0;
+      let upload = readFile();
+
+      $table.hide();
+
+      {{if $hasSMSConfig}}
+      if (typeof (FileReader) == "undefined") {
+        flash.error("Your browser does not support the required HTML5 file reader.");
+      } else {
+        $csv.prop('disabled', false);
+      }
+      {{end}}
+
+      $csv.change(function(file) {
+        let fileName = file.target.files[0].name;
+        $fileLabel.html(fileName);
+        $import.prop('disabled', false);
+      });
+
+      $cancel.on('click', function(event) {
+        upload.cancel();
+        flash.error('Canceled batch upload.');
+      });
+
+      $newCode.on('click', function(event) {
+        $retryCode.val(genCode());
+      });
+
+      $form.on('submit', function(event) {
+        event.preventDefault();
+        $import.prop('disabled', true);
+        $cancel.prop('disabled', false);
+
+        $table.show(100);
+        $progressDiv.removeClass('d-none');
+
+        if ($rememberCode.is(':checked')) {
+          setCookie('retryCode',$retryCode.val(),14);
+        } else {
+          setCookie('retryCode','',-1);
+        }
+
+        var reader = new FileReader();
+        reader.onload = upload.start;
+        reader.readAsText($csv[0].files[0]);
+      });
+
+      const batchSize = 1;
+
+      function readFile() {
+        let retryCode = $retryCode.val();
+        // State for managing cleanup and canceling
+        let cancelUpload = false;
+        let cancel = () => {
+          cancelUpload = true;
+        };
+
+        let start = async function(e) {
+          let rows = e.target.result.split('\n');
+          let batch = [];
+          totalUsersCreated = 0;
+          $tableBody.empty();
+          for (let i = parseInt($startAt.val()); i < rows.length && !cancelUpload; i++) {
+            if (rows[i].trim() != "") {
+              if (batch.length >= batchSize) {
+                $tableBody.empty();
+                batch = [];
+              }
+
+              let request = {};
+              let cols = rows[i].split(',');
+              request["phone"] = cols[0].trim();
+              request["symptomDate"] = (cols.length > 1) ? cols[1].trim() : "";
+              let hs  = String(CryptoJS.HmacSHA256(request["phone"], retryCode)).substr(0,36)
+              request["uuid"] = hs.substr(0,8)+ '-' + hs.substr(9,4) + '-' + hs.substr(13,4) + '-' + hs.substr(17,4) + '-' + hs.substr(21,12);
+
+              let row = "<tr><td>" + request["phone"] + "</td><td>" + request["symptomDate"] + "</td></tr>";
```

Comment:
This is vulnerable to XSS if the CSV contains HTML. Consider building these objects with jQuery and `.html` and `.text` instead.

Reply:
Isn't it more the server's responsibility to ensure escaping? Anyone could make a malicious client

Reply:
escaped anyway